File: neptune-streams-utils/examples/streams-to-firehose/stream_handler.py
# Copyright Amazon.com, Inc. or its affiliates.
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#    http://aws.amazon.com/apache2.0/
#
# or in the "license" file accompanying this file.
# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
# either express or implied. See the License for the specific language governing permissions
# and limitations under the License.

import json
import logging
import os
import boto3
import lambda_function
from commons import *
from handler import AbstractHandler, HandlerResponse

logger = logging.getLogger('StreamHandler')
logger.setLevel(logging.INFO)

class StreamHandler(AbstractHandler):

    def handle_records(self, stream_log):
        
        params = json.loads(os.environ['AdditionalParams'])
        delivery_stream_name = params['delivery_stream_name']
        
        client = boto3.client('firehose')

        records = stream_log[RECORDS_STR]
        
        last_op_num = None
        last_commit_num = None    
        count = 1
        
        firehose_records = []
        
        try:
            for record in records:
                
                # Process record
                if count % 500 == 0:
                    response = client.put_record_batch(
                        DeliveryStreamName=delivery_stream_name,
                        Records= firehose_records
                    )
                    logger.info(response)
                    logger.info(len(firehose_records))
                    firehose_records.clear()
                firehose_record = {
                    "Data": '{}\n'.format(json.dumps(record))
                }
                firehose_records.append(firehose_record)
                        
                # Update local checkpoint info
                last_op_num = record[EVENT_ID_STR][OP_NUM_STR]
                last_commit_num = record[EVENT_ID_STR][COMMIT_NUM_STR]
                count += 1
                
            if len(firehose_records) > 0:
                logger.info(len(firehose_records))
                response = client.put_record_batch(
                    DeliveryStreamName=delivery_stream_name,
                    Records=firehose_records
                )
                logger.info(response)
                   
            
        except Exception as e:
            logger.error('Error occurred - {}'.format(str(e)))
            raise e
        finally:
            try:
                yield HandlerResponse(last_op_num, last_commit_num, count)     
            except Exception as e:
                logger.error('Error occurred - {}'.format(str(e)))
                raise e

            
       
        
            
        

File: neptune-streams-utils/examples/java8/src/main/java/stream_handler/StreamHandler.java
/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License").
You may not use this file except in compliance with the License.
A copy of the License is located at
    http://www.apache.org/licenses/LICENSE-2.0
or in the "license" file accompanying this file. This file is distributed
on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
express or implied. See the License for the specific language governing
permissions and limitations under the License.
*/

package stream_handler;

import com.amazonaws.neptune.StreamsRecord;
import com.amazonaws.neptune.StreamsResponse;
import com.amazonaws.neptune.config.CredentialsConfig;
import org.apache.tinkerpop.gremlin.driver.Cluster;
import org.apache.tinkerpop.gremlin.driver.SigV4WebSocketChannelizer;
import org.apache.tinkerpop.gremlin.driver.remote.DriverRemoteConnection;
import org.apache.tinkerpop.gremlin.driver.ser.Serializers;
import org.apache.tinkerpop.gremlin.process.traversal.AnonymousTraversalSource;
import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
import utils.EnvironmentVariablesUtils;

import java.io.IOException;
import java.util.Map;

public class StreamHandler extends AbstractStreamHandler {

    private final GraphTraversalSource g;
    private final Cluster cluster;

    public StreamHandler(String neptuneEndpoint,
                         Integer neptunePort,
                         CredentialsConfig credentialsConfig,
                         Map<String, Object> additionalParams) {
        super(neptuneEndpoint, neptunePort, credentialsConfig, additionalParams);

        this.cluster = createCluster();
        this.g = AnonymousTraversalSource
                .traversal()
                .withRemote(DriverRemoteConnection.using(cluster));
    }

    @Override
    public void handleRecords(StreamsResponse streamsResponse) throws IOException {

        StreamsResponse.LastEventId lastEventId = new StreamsResponse.LastEventId();
        int recordCount = 0;

        for (StreamsRecord record : streamsResponse.getRecords()) {

            String op = record.getOp();

            if (op.equals("ADD")) {
                String id = record.getData().getId();
                String type = record.getData().getType();
                if (type.equals("vl")) {
                    System.out.println(g.V(id).valueMap(true).toList());
                } else if (type.equals("e")) {
                    System.out.println(g.E(id).valueMap(true).toList());
                }
            }

            StreamsRecord.EventId eventId = record.getEventId();

            lastEventId.setCommitNum(eventId.getCommitNum());
            lastEventId.setOpNum(record.getEventId().getOpNum());

            recordCount++;
        }

        streamsResponse.setLastEventId(lastEventId);
        streamsResponse.setTotalRecords(recordCount);

    }

    @Override
    public void close() throws IOException {
        cluster.close();
    }

    private Cluster createCluster() {
        Cluster.Builder builder = Cluster.build()
                .addContactPoint(String.valueOf(additionalParams.get("neptune_cluster_endpoint")))
                .port((int) additionalParams.get("neptune_port"))
                .enableSsl(true)
                .minConnectionPoolSize(1)
                .maxConnectionPoolSize(1)
                .serializer(Serializers.GRAPHBINARY_V1D0)
                .reconnectInterval(2000);

        if (Boolean.parseBoolean(EnvironmentVariablesUtils.getOptionalEnv("iam_auth_enabled", "false"))) {
            builder = builder.channelizer(SigV4WebSocketChannelizer.class);
        }

        return builder.create();
    }
}
